File: quicksort/lomuto/lomuto.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int partition(int* list, int low, int high);
void quickSort(int* list, int low, int high);
void swap(int* a, int* b);

int main(){
    int arr[] = {5,7,4,3,6,8,9,1,2,10};

    printf("before sort: ");
    for(int i = 0; i < 10; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");

    quickSort(arr, 0, 9);

    printf("after sort: ");
    for(int i = 0; i < 10; i++){
        printf("%d ", arr[i]);
    }
}

void quickSort(int* list, int low, int high){
    if(low < high){ // partition array and get pivot
        int pivot = partition(list, low, high);

        quickSort(list, low, pivot - 1); // sort low sub-array
        quickSort(list, pivot + 1, high); // sort high sub-array
    }
}

int partition(int* list, int low, int high){
    int pivot = list[high]; // set last elem to be pivot
    int i = low - 1;

    for(int j = low; j <= high-1; j++){
        if(list[j] <= pivot){
            i++;
            swap(&list[i], &list[j]);
        }
    }

    swap(&list[i+1], &list[high]); // swap pivot with where i is
    return i+1;
}

void swap(int* a, int* b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

File: quicksort/lomuto/lomuto.py
def QuickSort(array, low, high):
    if low < high:
        pivot = partition(array, low, high)     # find pivot
        QuickSort(array, low, pivot-1)          # sort low sub-array
        QuickSort(array, pivot+1, high)         # sort high sub-array

def partition(array, low, high):
    pivot = array[high]
    i = low - 1
    
    for j in range(low,high):
        if array[j] <= pivot:
            i += 1
            array[j], array[i] = array[i], array[j]     # swap
    array[i+1], array[high] = array[high], array[i+1]   # swap
    return i+1


test_list = [10,9,4,5,3,2,6,7,8,1]
QuickSort(test_list, 0, len(test_list)-1)
print(test_list)

